### ISSUE TITLE
Thumbnails are not loading on start of a new browser session for new tabs

### ISSUE THREAD

**Issue by Bundaboy53** [fastaddons/GroupSpeedDial#88]
**Describe the bug**
Thumbnails are not loading on start of a new browser session for new tabs and will continue to behave like that for the entire session unless the extension is disabled and then enabled. This clears the problem until the next session.

**Steps To Reproduce**
Start a new session (ie start FF), attempt to open a new tab. Observe that thumbnails are not loading.

**Expected behavior**
All thumbnails loaded for each dial

**Device info:**
 - operating system: W7 
 - Browser and version: FF 94.0.2 (32-bit)
 - GroupSpeedDial version: 16.8

**Additional info**
Problem started with the release of 16.8 - I see that there is a newer release (16.9) which should have updated automatically but hasn't. I will investigate that to see if the problem is fixed in 16.9 and report back. EDIT [Ok, I have updated to 16.9, not sure why that didn't update automatically, but the problem seems to be fixed. I will report back IF it returns]

**Comment by fastaddons** [comment]
Could you send me a screenshot of those not-loaded thumbnails?
The 16.9 has only some minor changes, but also 16.8 didn't changed thumbnails loading or handling.
Could you also try to visit this storage test page?
https://firefox-storage-test.glitch.me/
(the test takes about 10 seconds and you should see "Storage is working." message)

**Comment by Bundaboy53** [comment]
> Could you send me a screenshot of those not-loaded thumbnails? The 16.9 has only some minor changes, but also 16.8 didn't changed thumbnails loading or handling. Could you also try to visit this storage test page? https://firefox-storage-test.glitch.me/ (the test takes about 10 seconds and you should see "Storage is working." message) EDIT: I will check it tomorrow, now it's time to sleep :)

I have updated to 16.9 and the problem seems to be fixed. If it returns I will post a screenshot - I was going to do that and forgot, tch.
Sleep well!

**Comment by fastaddons** [comment]
I'm glad hear that!
Have a nice upcoming weekend :)

**Comment by Bundaboy53** [comment]
Unfortunately, when I opened FF this morning, the problem had returned. I attach a screen grab of the issue. To be honest I think this is more of a FF issue than a SpeedDial issue, but it's hard to tell for sure. I note if I close FF and reopen it the problem goes away, so it's not a major issue
![Capture](https://user-images.githubusercontent.com/60862216/143657744-2ba55f54-4052-4eb6-a9cb-88982b3ff78b.JPG)
e.

**Comment by fastaddons** [comment]
Is there also some error message in the corner?
If not, could you please send me the runtime logs? You can find it in the main menu / Advanced / Send error logs. Copy and paste it here or send it to my email.

**Comment by Bundaboy53** [comment]
Yah! I cannot now reproduce the problem at all - don't you hate that?
Just for the record, I see no error messages, and I have sent the log as requested (although it doesn't indicate any errors), but as I can't reproduce the problem that's not surprising.
I don't think you should waste anymore time on this, if I get something concrete for you to go on I'll be back in touch, in the meantime it's an easy enough work around even if the problem does recur. Thanks for your time and assistance. Gary.

**Comment by fastaddons** [comment]
Hello, again,
Thank you for the error logs. Even though there are no errors, there are time stamps it and says that it took 36 seconds to start the background script and open the database with thumbnails. That's not good.
So something is definitely wrong because these operations should take less than a second.

One thing you can try is backup your dials and thumbnails to a file (export them from the Options page / Import backup tab) and then reinstall my addon - you have to restart Firefox after uninstalling. Then when you install it back, it should be completely empty and when you import your data back it should all look the same as before.

**Comment by Bundaboy53** [comment]
Thanks Juraj.
I had the problem again this morning. A possible explanation for the long delay in starting is that I have FF starting up at boot time when there is a great deal of activity going on, everything starts slowly on my computer during that time.
I have emailed another log report taken when the issue is occurring but it doesn't seem to show anything further.

I will try backup reinstall process and get back to you. Cheers.

**Comment by i-3Dmax** [comment]
Сегодня заметил, что это дополнение перестало работать. Миниатюры не отображаются. В настройки не возможно зайти. ![GIF 28 11 2021 13-57-59](https://user-images.githubusercontent.com/95168050/143765220-ee1ff031-4162-4bb8-92db-5412f9056597.gif)

**Comment by fastaddons** [comment]
@i-3Dmax if not even dials loads, that usually means the storage is not accessible at all. But usually restarting Firefox or whole PC fixes it. Please let me know if you see it also after restarting PC. This error is very rare and it may pointing to a problem with your Firefox or drive. You can also check the storage test page to see if your Firefox storage engine works fine:
https://firefox-storage-test.glitch.me/
(the test takes about 10 seconds and you should see "Storage is working." message)

@Bundaboy53 the slow start when you turn on PC does explain the longer than usual loading times. But still it feels way too long time to load only few KB of data. I hope reinstalling will help. Another thing you could try is Firefox Reset - that can fix Firefox related issues by recreating you profile and migrating your passwords and bookmarks (but not addons data, so before you try that, make sure to backup any addons data). 
https://support.mozilla.org/kb/refresh-firefox-reset-add-ons-and-settings#w_refresh-firefox

As a bonus tip, if this happens after turning ON PC, one workaround would be to use Hibernation or Sleep instead of Shut-down. Using sleep will keep all your apps opened and your PC will wake up into fully ready state in few seconds. I use sleep every day and restart PC only when it needs to install updates.

**Comment by i-3Dmax** [comment]
После перезагрузки ПК никаких изменений. Дополнение по прежнему не работает. С другими дополнениями вроде всё впорядке
![2021-11-28_155421](https://user-images.githubusercontent.com/95168050/143768650-9eecddea-d455-421c-9cc8-5b99bffc3325.png)
.

**Comment by fastaddons** [comment]
@i-3Dmax could you send me your error logs? Thanks!
![_send_error_logs](https://user-images.githubusercontent.com/55486931/143769136-aa519e87-0f39-4eef-b082-cd93bdd584b4.png)

**Comment by i-3Dmax** [comment]
Не возможно
![GIF 28 11 2021 16-14-51](https://user-images.githubusercontent.com/95168050/143769465-ef9734d8-2d24-4502-be7a-73600aea763e.gif)

**Comment by fastaddons** [comment]
@i-3Dmax wow, that must be completely broken.
Cold you send me screnshot of the developer console? Open the broken dials page and press "Ctrl + Shift + K".

**Comment by i-3Dmax** [comment]
![2021-11-28_162210](https://user-images.githubusercontent.com/95168050/143769645-d6a727b5-467d-4830-9421-5bf6d7f99ffb.png)

**Comment by fastaddons** [comment]
Could you enable the error logging?
![image](https://user-images.githubusercontent.com/55486931/143769773-071afd72-7cf4-4b41-b36c-e2c0b2be91a6.png)

(the first option needs to be checked and I see there is 1 error hidden)
Enable it and reload the page. Thanks!

**Comment by i-3Dmax** [comment]
![2021-11-28_162900](https://user-images.githubusercontent.com/95168050/143769891-2f5a0442-f5a5-491a-a335-5e3ec290ca49.png)

**Comment by i-3Dmax** [comment]
Alas! There is no backup. How can you do it in this situation?

**Comment by fastaddons** [comment]
There are some automatic backups but just in case, let's make few additional backups first.
This is what you do:
1) open dials page
2) open console with "Ctrl + Shift + K"
3) write there this and press enter:
```
JSON.stringify(await browser.storage.local.get())
```
4) copy the result and paste it into empty text file (use Notepad for example)
![Screenshot 2021-11-28 14 47 54](https://user-images.githubusercontent.com/55486931/143770613-e124f7c3-dc2a-4c77-baf0-8ca1d72166cb.png)
(make sure to click Copy object, not Copy message)

Then repeat the step 3 with following command:
```
JSON.stringify(await browser.storage.sync.get())
```
And then one more time with this command:
```
JSON.stringify(localStorage)
```
Hopefully you will be able to execute these without errors.

**Comment by i-3Dmax** [comment]
Сделал всё по иструкции. Потом удалил дополнение, перезагрузил браузер, установил. Но оно так и не работает (((
![GIF 28 11 2021 17-13-46](https://user-images.githubusercontent.com/95168050/143771629-fcbfcd28-a7bf-4fc8-8a6d-96f96ebc5051.gif)

**Comment by i-3Dmax** [comment]
Настройки дополнения кажется тоже не работают
![GIF 28 11 2021 17-19-16](https://user-images.githubusercontent.com/95168050/143771815-aeba9527-c086-49eb-900e-469312f8efe6.gif)
.

**Comment by fastaddons** [comment]
Yeah the Options page won't work either if the dials page won't load.
Also, do you use some special settings in your Firefox? Something like disable history, or some restricted privacy options?
about:preferences#privacy
Do you have it set to "Standard"?

And could try to install some other similar addon to see if it will work?
And lastly, try to uninstall it again and restart whole PC before installing it back (to be be sure all firefox.exe processes were killed).

**Comment by i-3Dmax** [comment]
The privacy setting is Standard.
I deleted the add-on, rebooted the PC, installed it again - everything is the same, it didn't help.
I tried adding Quick Dial and Speed Dial 2 - they working fine.

**Comment by fastaddons** [comment]
Oh man, this makes no sense.
Let's do a "sanity check" :)
1) open page about:profiles
2) click "Create new profile"
3) give it some name, like "testing"
4) click the "Launch profile in new browser"
That will open brand new empty Firefox - try to install my addon there. Does it work?
After you are done, click the "Remove" button to remove the testing profile.

If it will work in testing profile, then we know this must be related with something broken in your current profile. And that can be fixed with [Firefox Reset ]( https://support.mozilla.org/kb/refresh-firefox-reset-add-ons-and-settings#w_refresh-firefox)(which also creates new account and then migrates all passwords and bookmarks, but no addons data).

**Comment by i-3Dmax** [comment]
Да, да, да )))))
Тоже об этом подумал ))
С новым профилем дополнение заработало ))

**Comment by i-3Dmax** [comment]
Thank you very much! The extension has been restored!!! And now how to restore the dials?

**Comment by i-3Dmax** [comment]
It turned out that the add-on found automatically created fresh backups. One was a week old. Everything is restored! Thank you very much!

**Comment by fastaddons** [comment]
Hello,
I'm sorry for the delay.
I could give you instructions to restore the backup we made today if you need.

**Comment by i-3Dmax** [comment]
Yes, that would be great! Otherwise, the restored backup is not entirely correct.

**Comment by fastaddons** [comment]
So it should be simple, you need to restore only one if the backups, the one created with `JSON.stringify(await browser.storage.local.get())` command.
Also, before you start, export your data to a file from the Options page / Import backup tab.

So, to restore it, open Dials page, open console (Ctrl + Shift + K) and write there:
```
browser.storage.local.set(JSON.parse(`
// instead of this line, paste here the backup text
`.trim()))
```
And press enter that that should be it.

**Comment by i-3Dmax** [comment]
An error occurred while executing this request:

Uncaught SyntaxError: JSON.parse: unexpected character at line 1 column 2 of the JSON data
    <anonymous> debugger eval code:1
debugger eval code:1:32
    <анонимный> debugger eval code:1
![2021-11-29_191743](https://user-images.githubusercontent.com/95168050/143903849-ba58e720-a7d5-4dce-adbb-434903376ab3.png)

**Comment by fastaddons** [comment]
Could you post here first 10 and last 10 characters of the backup?

**Comment by i-3Dmax** [comment]
Request made:
browser.storage.local.set(JSON.parse(
{"_error_getImages":2,"_forceCallback":1
...
"tutorial":1,"user":null,"version":xxxxxxxxxxxxx}
))

**Comment by i-3Dmax** [comment]
Super! Everything worked out!!! Thank you so much for your work !!!

**Comment by fastaddons** [comment]
I'm happy to hear that! And I'm sorry for the confusion with restoring the backup, this kind of emergency recovery is very rare :).
Have a nice day  :)
